pyxem/signals/segments.py

```diff
@@ -0,0 +1,443 @@
+# along with pyXem.  If not, see <http://www.gnu.org/licenses/>.
+"""Signal classes for segmentation results obtained from
+i) machine learning (LearningSegment) and ii) VDF images (VDFSegment).
+
+"""
+
+import numpy as np
+from tqdm import tqdm
+
+from hyperspy.signals import Signal2D
+
+from pyxem.utils.segment_utils import (norm_cross_corr, separate_watershed,
+                                       get_gaussian2d)
+from pyxem.signals.diffraction_vectors import DiffractionVectors
+from pyxem.signals.electron_diffraction2d import ElectronDiffraction2D
+from pyxem.signals import transfer_signal_axes
+
+from traits.trait_errors import TraitError
+
+
+class LearningSegment:
+    _signal_type = "learning_segment"
+
+    def __init__(self, factors, loadings):
+        # Factors as Signal2D
+        self.factors = factors
+        # Corresponding loadings as Signal2D
+        self.loadings = loadings
+
+    def correlate_learning_segments(self, corr_th_factors=0.4,
+                                    corr_th_loadings=0.4):
+        """Iterates through the factors and loadings and calculates the
+        normalized cross-correlation between all factors and all
+        loadings. Factors and loadings are summed if the correlations of
+        both factors and loadings exceed the given thresholds.
+
+        Parameters
+        ----------
+        corr_th_factors : float
+            Correlation threshold for factors. Must be between -1 and 1.
+            Factors and loadings are summed if both factors and loadings
+            have normalized cross-correlations above corr_th_factors and
+            corr_th_loadings, respectively.
+        corr_th_loadings : int, optional
+            Correlation threshold for loadings. Must be between -1 and 1.
+
+        Returns
+        -------
+        learning_segment : LearningSegment
+            LearningSegment where possibly some factors and loadings
+            have been summed.
+        """
+        # If a mask was used during the decomposition, the factors and/or
+        # loadings will contain nan, which must be converted to numbers prior
+        # to the correlations calculations.
+        factors = self.factors.map(lambda x: np.nan_to_num(x), inplace=False)
+        loadings = self.loadings.map(lambda x: np.nan_to_num(x), inplace=False)
+        factors = factors.copy().data
+        loadings = loadings.copy().data
+        correlated_loadings = np.zeros_like(loadings[:1])
+        correlated_factors = np.zeros_like(factors[:1])
+
+        # For each loading and factor, calculate the normalized
+        # cross-correlation to all other loadings and factors, and define
+        # add_indices for those with a value above corr_th_loadings and
+        # corr_th_factors respectively.
+        while np.shape(loadings)[0] > 0:
+            corr_list_loadings = list(map(
+                lambda x: norm_cross_corr(x, template=loadings[0]), loadings))
```

Comment:
a lambda inside a map doesn't seem like a good thing - why is this not a hyperspy.Signal.map() in any case?

Reply:
Maybe I am missing something, but I don't get it to work with hyperspy's map:        `corr_list_loadings = loadings.map(norm_cross_corr, inplace=False, template=loadings.data[0])` Do you know what I am doing wrong?

Reply:
shot in the dark (ie- I didn't run the code) but `template` is an argument of `norm_cross_corr` so need to come before `inplace` in the argument order.  In general it goes:  `.map(function, function_args, map_args)`

Reply:
Will implement (as check)

Reply:
This is a larger refactor than originally thought. I am going to leave it for now.